src/eddymotion/model/dmri_utils.py

```diff
@@ -0,0 +1,129 @@
+    shell_data : ndarray
+        Volumes corresponding to the given ``bvals``.
+    output_bvals : ndarray
+        Selected b-values (as extracted from ``bvals``).
+    output_bvecs : ndarray
+        Selected b-vectors.
+    """
+
+    indices = [get_bval_indices(bvals, shell, tol=tol) for shell in bvals_to_extract]
+    indices = np.unique(np.sort(np.hstack(indices)))
+
+    if len(indices) == 0:
+        raise ValueError(
+            f"No DWI volumes found corresponding to the given b-values: {bvals_to_extract}"
+        )
+
+    shell_data = dwi.get_fdata()[..., indices]
+    output_bvals = bvals[indices].astype(int)
+    output_bvecs = bvecs[indices, :]
+
+    return indices, shell_data, output_bvals, output_bvecs
+
+
+def find_shelling_scheme(bvals, tol=SHELL_DIFF_THRES):
+    """Find the shelling scheme on the given b-values: extract the b-value
+    shells as the b-values centroids using k-means clustering.
+
+    Parameters
+    ----------
+    bvals : :obj:`ndarray`
+         b-values in FSL format.
+    tol : :obj:`int`, optional
+        Tolerance between the b-values and the centroids in the average squared
+        distance sense.
+
+    Returns
+    -------
+    shells : :obj:`ndarray`
+        b-value shells.
+    bval_centroids : :obj:`ndarray`
+        Shell value corresponding to each value in ``bvals``.
+    """
+
+    # Use kmeans to find the shelling scheme
+    for k in range(1, len(np.unique(bvals)) + 1):
+        kmeans_res = KMeans(n_clusters=k).fit(bvals.reshape(-1, 1))
+        # ToDo
+        # The tolerance is not a very intuitive value, as it has to do with the
+        # sum of squared distances across all samples to the centroids
+        # (_inertia)
+        # Alternatives:
+        # - We could accept the number of clusters as a parameter and do
+        # kmeans_res = KMeans(n_clusters=n_clusters)
+        # Setting that to 3 in the last testing case, where tol = 60 is not
+        # intuitive would give the expected 6, 1000, 2000 clusters.
+        # Passes all tests. But maybe not tested corner cases
+        # We could have both k and tol as optional parameters, set to None by
+        # default to force the user set one
+        # - Use get_bval_indices to get the cluster centroids and then
+        # substitute the values in bvals with the corresponding values
+        # indices = [get_bval_indices(bvals, shell, tol=tol) for shell in bvals_to_extract]
+        # result = np.zeros_like(bvals)
+        # for i, idx in enumerate(indices):
+        #   result[idx] = bvals_to_extract[i]
+
+        if kmeans_res.inertia_ / len(bvals) < tol:
+            break
+    else:
+        raise ValueError(f"bvals parsing failed: no shells found more than {tol} apart")
+
+    # Convert the kclust labels to an array
+    shells = kmeans_res.cluster_centers_
+    bval_centroids = np.zeros(bvals.shape)
+    for i in range(shells.size):
+        bval_centroids[kmeans_res.labels_ == i] = shells[i][0]
```

Comment:
Sorry, was not aware of that code snippet. If that does a better job than the one proposed here, happy to change. I have tried it locally and looks to perform well.

> That said, perhaps a more straightforward way of doing this is to just bin the b-values into, say, 15 bins with a range of b-values from 0 through, say, 8,000. Then, pick nonempty bins, and if they are below seven, call the dataset multi-shell, and if they are above seven, call them DSI.

Not sure if I see this: are you saying that the `mrqc` snippet should be edited (assuming we C&P here) following what you are describing here? Do you think that the current `mriqc` implementation will be more robust with those changes? Unless the `mriqc` implementation is flawed, I would leave that for a separate PR. We can open an issue once this is merged.

Reply:
> Sorry, was not aware of that code snippet. If that does a better job than the one proposed here, happy to change. I have tried it locally and looks to perform well.

What works well? The mriqc snippet, this snippet or both :D



> Not sure if I see this: are you saying that the `mrqc` snippet should be edited (assuming we C&P here) following what you are describing here?

Yes, my impression is that k-means is an overkill for this problem. Just calculating the histogram with the adequate number of bins and min/max definitions should suffice.

So, instead of C&P, I guess we could try the solution here en eddymotion and later propagate into mriqc if it is successful.